tinygrad/codegen/uops.py

```diff
@@ -307,30 +315,22 @@ def print(self):
     for i,u in enumerate(self):
       print(f"{i:4d} {str(u.op):20s}: {str(u.dtype) if u.dtype is not None else '':25s} " f"{str([self.uops.index(x) for x in u.src]):32s} {u.arg}")
 
-  def graph_rewrite(self, sink:UOp, pm:PatternMatcher):
-    # recursive rewrite
-    changed = getenv("UOPS_REWRITE", 1)
+  @functools.lru_cache(None)  # pylint: disable=method-cache-max-size-none
+  def recursive_rewrite(self, pm: PatternMatcher, up:UOp) -> UOp:
+    recurse_cnt = 0
+    while (rewritten:=pm.rewrite(up)) is not None:
+      assert recurse_cnt < 100, f"recursive_rewrite looped {up} <--> {rewritten}"
+      up = rewritten
+      recurse_cnt += 1
+    up = UOp(up.op, up.dtype, tuple(self.recursive_rewrite(pm, x) for x in up.src), up.arg)
```

Comment:
Create a new one, no need to clear the caches. This change also allows UOp to become a frozen dataclass (which is a tad slower though).